### PR TITLE
Switch how dotnet lambda debugger works for SAM 1.0

### PR DESCRIPTION
**Pull request by abrooksv**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/SpinUtils.kt [hunk 1]
```diff
@@ -20,3 +20,18 @@ fun spinUntil(duration: Duration, condition: () -> Boolean) {
         }
     }
 }
+
+/**
+ * Keeps running the function until it returns a non-null value. Checks every 100ms
+ */
+suspend fun <T> spinUntilResult(duration: Duration, func: () -> T?): T {
```
Comment (rli) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/utils/SpinUtils.kt:27]: looks familiar, does this appear elsewhere, or am i thinking of another codebase?

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt [hunk 2]
```diff
@@ -24,60 +31,47 @@ import com.jetbrains.rd.util.put
...
- *      '$ dotnet <debug_args_list> MockBootstraps.dll --debugger-spin-wait'
- *
- * We use <debugger_args_list> to inject a simple custom dotnet core application 'JetBrains.Rider.Debugger.Launcher'
- * that run Rider's debugger. This app starts Rider's debugger under Mono runtime that is bundled with Rider SDK (runtime.sh).
- * The app read arguments and start a Rider's DebuggerWorker process with ports and compiled lambda DLL - MockBootstraps.dll
- *
- * 'JetBrains.Rider.Debugger.Launcher' app is hosted on github repo - https://github.com/JetBrains/JetBrains.Rider.Debugger.Launcher.
- * The app is packed to NuGet package and is accessible in Rider SDK in '<SDKPath>/lib/ReSharperHost/' directory.
+ * Workflow:
+ * 1. Find the correct container by filtering `doker ps` for one of the published port of the debugger
```
Comment (rli) [inline jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:64]: nit: `doker` -> `docker`

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt [hunk 3]
```diff
@@ -24,60 +31,47 @@ import com.jetbrains.rd.util.put
...
+ * 2. Use `docker exec` and pgrep to locate the dotnet PID of the runtime.
```
Comment (rli) [inline jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:65]: are we guaranteed to have pgrep? iirc it's non-standard, even if though it's frequently available

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt [hunk 4]
```diff
@@ -24,60 +31,47 @@ import com.jetbrains.rd.util.put
...
  *
- * Debugger is launched via SAM CLI command: '$ sam local invoke' and pass debugger parameters --debugger-path and --debugger-args.
- * Under SAM CLI this command is translated into:
- *    - volume that is mounted in Docker to launch debugger (--debugger-path).
- *      This volume is set to path with debugger process used in Rider - DebuggerWorker - inside Rider SDK.
- *    - additional debugger arguments that are applied when launching debugger inside Docker (--debugger-args):
- *      '$ dotnet <debug_args_list> MockBootstraps.dll --debugger-spin-wait'
- *
- * We use <debugger_args_list> to inject a simple custom dotnet core application 'JetBrains.Rider.Debugger.Launcher'
- * that run Rider's debugger. This app starts Rider's debugger under Mono runtime that is bundled with Rider SDK (runtime.sh).
```
Comment (rli) [inline jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:66]: what's the change that removes the need to run under the bundled mono runtime?

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt [hunk 5]
```diff
@@ -139,118 +113,117 @@ class DotNetSamDebugSupport : SamDebugSupport {
...
+        val workerModel = RiderDebuggerWorkerModelManager.createDebuggerModel(debuggerLifetime, protocol)
+
         val executionResult = state.execute(environment.executor, environment.runner)
+        val console = TextConsoleBuilderFactory.getInstance().createBuilder(environment.project).console
+        val samProcessHandle = executionResult.processHandler
+        console.attachToProcess(samProcessHandle)
+
+        // If we have not started the process's notification system, start it now.
+        // This is needed to pipe the SAM output to the Console view of the debugger panel
+        if (!samProcessHandle.isStartNotified) {
```
Comment (rli) [inline jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:144]: what's the case where the process has already started?

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt (hunk 2 above)
Comment (abrooksv) [reply jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:64]: Fixed

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt [hunk 6]
```diff
@@ -261,16 +234,44 @@ class DotNetSamDebugSupport : SamDebugSupport {
...
+    }
+
+    private suspend fun findDotnetPid(dockerContainer: String): Int = spinUntilResult(Duration.ofSeconds(30)) {
+        ExecUtil.execAndGetOutput(
+            GeneralCommandLine(
+                "docker",
+                "exec",
+                "-i",
+                dockerContainer,
+                "/usr/bin/pgrep",
```
Comment (hunterwerlla) [inline jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:256]: This won't work on windows?

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:65]: Good question... I'll see what I can find out

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt (hunk 4 above)
Comment (abrooksv) [reply jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:66]: Starting in 193, the debugger can run in dotnet core. We use it for cloud debug in that mode too

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt (hunk 5 above)
Comment (abrooksv) [reply jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:144]: Doesn't hurt. It is already pumping events out to the process listeners.

Preventing the call prevents a LOG.error() that gets invoked if the AtomicReference is already set to started state.

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt (hunk 6 above)
Comment (abrooksv) [reply jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:256]: This is executed inside of the container which is Amazon Linux 2 based

File: jetbrains-core/src/software/aws/toolkits/jetbrains/utils/SpinUtils.kt (hunk 1 above)
Comment (abrooksv) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/utils/SpinUtils.kt:27]: Yaaaa, its similar to software/aws/toolkits/core/utils/Waiter.kt:51

File: jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt (hunk 3 above)
Comment (abrooksv) [reply jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt:65]: If we can't, we will go with this beauty:

```
find /proc -mindepth 2 -maxdepth 2 -name exe -exec ls -l {} \; 2>/dev/null | grep dotnet | sed -n 's/.*\/proc\/\(.*\)\/exe.*/\1/p'
```

**Comment by hunterwerlla** [comment]
Looks good to me, tested on my mac with SAM 0.47.0, was able to run and debug. Will approve once the pgrep change is made

**Comment by abrooksv** [comment]
> Looks good to me, tested on my mac with SAM 0.47.0, was able to run and debug. Will approve once the pgrep change is made

It was changed, I forgot I pushed it.

Its now using the long "find" command

**Pull request merged**